core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -184,13 +194,43 @@ public void onRemoval(DnsQuestion question, @Nullable List<DnsRecord> records,
 
         assert question instanceof DnsQuestionWithoutTrailingDot;
         final DnsQuestionWithoutTrailingDot cast = (DnsQuestionWithoutTrailingDot) question;
-        final CacheEntry entry = addressResolverCache.getIfPresent(cast.originalName());
-        if (entry != null && entry.refreshable()) {
-            // onRemoval is invoked by the executor of 'dnsResolverCache'.
-            executor().execute(entry::refresh);
+        final String hostname = cast.originalName();
+
+        if (!autoRefresh) {
+            addressResolverCache.invalidate(hostname);
+            return;
+        }
+
+        final CacheEntry entry = addressResolverCache.getIfPresent(hostname);
+        if (entry != null) {
+            if (entry.refreshable()) {
+                // onRemoval is invoked by the executor of 'dnsResolverCache'.
+                executor().execute(entry::refresh);
+            } else {
+                // Remove the old CacheEntry.
+                addressResolverCache.invalidate(hostname);
+            }
         }
     }
 
+    @Override
+    public void onEviction(DnsQuestion question, @Nullable List<DnsRecord> records,
+                           @Nullable UnknownHostException cause) {
+        if (resolverClosed) {
+            return;
+        }
+
+        final DnsRecordType type = question.type();
+        if (!(type.equals(DnsRecordType.A) || type.equals(DnsRecordType.AAAA))) {
+            // AddressResolver only uses A or AAAA.
+            return;
+        }
+
+        final DnsQuestionWithoutTrailingDot cast = (DnsQuestionWithoutTrailingDot) question;
+        // The DnsCache is full. Don't schedule refreshing because it may cause another eviction.
```

Comment:
If users also specify the DnsCache ttl (not dns ttl, but caffeine ttl) I think this callback will be invoked.
This could potentially lead to increased latency due to intermittent dns queries whenever an entry is evicted.
I'm wondering if it's worth narrowing the scope of this callback to just `RemovalCause.SIZE`

Reply:
Nice idea. I totally agree with your idea.